### PR TITLE
HD Wallets support

### PR DESCRIPTION
**Pull request by decentraliser**

### CODE REVIEW

File: src/commands/profile/decrypt.ts [hunk 1]
```diff
@@ -34,8 +36,19 @@ export default class extends ProfileCommand {
...
         const password = await new PasswordResolver().resolve(options)
-        const account = profile.decrypt(password)
-        const text = new AccountCredentialsTable(account).toString()
-        console.log(text)
+
+        if (profile.type === 'PrivateKey') {
+            const account = profile.decrypt(password)
+            console.log(AccountCredentialsTable.createFromAccount(account).toString())
+        } else {
+            if (!profile.encryptedPassphrase || !profile.pathNumber) {return }
```
Comment (dgarcia360) [inline src/commands/profile/decrypt.ts:44]: In this case, the command could  print a message instead of returning an empty response.

Comment (decentraliser) [reply src/commands/profile/decrypt.ts:44]: It's just a trick for typing, this case shouldn't happen

Will modify it

**Comment by decentraliser** [comment]
I propose to do the derivation command in another PR if you are OK with that @dgarcia360 
I think this one is big enough

File: src/services/encryption.service.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
+/*
```
[the remark this answers is not in the document]
Comment (decentraliser) [reply src/services/encryption.service.ts:1]: I propose to keep it temporarily until a decision is made about whether or not the clients should keep encrypted mnemonic

Currently, there is no method available to encrypt/decrypt given by a symbol library, so there are no many choices

File: test/models/networkCurrency.model.spec.ts [hunk 3]
```diff
@@ -67,12 +67,11 @@ describe('Network currency', () => {
  })
 
  it('createRelative should throw when provided an invalid number', () => {
+  expect(() => mockNetworkCurrency.createRelative('not an amount')).to.throw()
   // @ts-ignore
```
[the remark this answers is not in the document]
Comment (decentraliser) [reply test/models/networkCurrency.model.spec.ts:71]: no, null and undefined are not types that are accepted by the function

However, during runtime, null and undefined could be passed to it

File: src/interfaces/create.profile.command.ts [hunk 4]
```diff
@@ -60,47 +104,35 @@ export class AccountCredentialsTable {
...
-                this.setDefaultProfile(simpleWallet.name)
-            }
+    protected createProfile(args: ProfileCreation): Profile {
+        try {
+            const profile = ProfileService.createNewProfile(args)
+            this.profileRepository.save(profile)
+            if (args.isDefault) {this.setDefaultProfile(args.name)}
             return profile
+        } catch (ignored) {
+            console.log('CreateProfileCommand -> createProfile -> ignored', ignored)
```
[the remark this answers is not in the document]
Comment (decentraliser) [reply src/interfaces/create.profile.command.ts]: nice catch, I forgot it

File: src/respositories/profile.repository.ts [hunk 5]
```diff
@@ -205,20 +192,18 @@ export class ProfileRepository {
...
-                    const migrations = Object.entries(allMigrations)
-                        .filter(([version]) => Number(version) > profileVersion)
-                        .map(([, migration]) => migration)
-
-                    // return migrated profiles
-                    return migrations
-                        .map((migration) => migration({[name]: profile}))
-                        .reduce((acc, migratedProfile) => ({...acc, ...migratedProfile}), {})
-                }
+                // Skip if profile is up-to-tate (shouldn't happen)
```
[the remark this answers is not in the document]
Comment (decentraliser) [reply src/respositories/profile.repository.ts:195]: Remove the shouldn't happen

There should be no case when the profiles do not all have the same version 
(except if someone edited the symbol-rc file without using the CLI)

File: src/services/encryption.service.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (evias) [reply src/services/encryption.service.ts:1]: I would confirm that this class _must not be added to the CLI_. There is already a class provided by HD wallet to encrypt if needed.

File: src/services/profile.service.ts [hunk 6]
```diff
@@ -83,4 +68,29 @@ export class ProfileService {
...
+        return profile
+    }
+
+    /**
+     * Creates a new profile from a DTO
+     * @static
+     * @param {ProfileDTO} args
+     * @returns {Profile}
+     */
+    public static createProfileFromDTO(args: ProfileDTO): Profile {
```
[the remark this answers is not in the document]
Comment (decentraliser) [reply src/services/profile.service.ts:92]: For our case, we want to take many `profileDTO`, and then create `PrivateKeyProfile `or `HdProfile `according to their properties

Then the choice of instantiating a type of profile or another has to be done before the instantiation of a profile subclass instantiation. So this function can not go into the Profile class.
(having the factory inside of the `Profile `class would create a circular dependency issue)

Another solution would be to create a `ProfileFactory `class, but I think it is more lightweight to put it in the profile service.

**Pull request merged**